engine/mapedit/objq.3
.\"	$Csoft$
.\"
.\" Copyright (c) 2002 CubeSoft Communications, Inc.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistribution of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Neither the name of CubeSoft Communications, nor the names of its
.\"    contributors may be used to endorse or promote products derived from
.\"    this software without specific prior written permission.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
.\" INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
.\" (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
.\" SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
.\" STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
.\" IN ANY WAY OUT OF THE USE OF THIS SOFTWARE EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd August 20, 2002
.Dt OBJQ 3
.Os
.Sh NAME
.Nm objq
.Nd agar object queue widget
.Sh SYNOPSIS
.Bd -literal
#include <engine/mapedit/mapedit.h>
#include <engine/mapedit/objq.h>
.Ed
.Sh DESCRIPTION
The
.Nm
widget displays a list of objects available for edition.
.Sh INITIALIZATION
.nr nS 1
.Ft struct objq *
.Fn objq_new "struct region *reg" "struct mapedit *med" "int flags" "int rw" "int rh"
.Pp
.Ft void
.Fn objq_init "struct objq *oq" "struct mapedit *med" "struct map *m" "int flags" "int rw" "int rh"
.nr nS 0
.Pp
The
.Fn objq_new
function allocates, initializes, and attaches a
.Nm
widget.
The
.Fn objq_init
function initializes a
.Nm
widget of size
.Fa rw
x
.Fa rh .
The
.Fa flags
cannot be changed at runtime, and may include:
.Pp
.Bd -literal
#define OBJQ_HORIZ		0x01	/* Display icons horizontally */
#define OBJQ_VERT		0x02	/* Display icons vertically */
.Ed
.Sh EVENTS
The
.Nm
widget reacts to the following events:
.Pp
.Bl -tag -compact -width 25n -indent
.It widget-scaled
Adjust the list display.
.It window-mousemotion
Right button scrolls the list.
.It window-mousebuttonup
Right button stops scrolling the list.
.El
.Pp
The
.Nm
widget does not generate any event.
.Sh SEE ALSO
.Xr agar 3 ,
.Xr mapview 3 ,
.Xr tilestack 3
.Sh HISTORY
The
.Nm
widget first appeared in AGAR 1.0.


CHANGELOG.md
# Changelog
All notable changes to Agar will be documented in this file. The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.8.0] - Unreleased
### Added
- [**Style Editor**](https://libagar.org/man3/AG_StyleEditor): Display computed style rules in addition to per-instance overrides.
- [**AG_Button**](https://libagar.org/man3/AG_Button): New style attributes `text-align` and `vertical-align`.
- [**AG_Widget**](https://libagar.org/man3/AG_Widget): Introduce the `style()` method, called by the style compiler to parse widget-specific style attributes into an efficient binary representation.
- [**AG_Object**](https://libagar.org/man3/AG_Object): New function `AG_AccessVariable_NoLock()` (lockless variant of `AG_AccessVariable()`).
- [**AG_Object**](https://libagar.org/man3/AG_Object): Document the `AG_OBJECT_BOUND_EVENTS` flag.

### Removed
- [**AG_Window**](https://libagar.org/man3/AG_Window): Remove the "window-detached" event. It is redundant since the generic "detached" event does the same job.

### Changed
- Major performance improvements in the style compiler. Since it runs whenever widgets are resized, this should lead to better fluidity and overall lower CPU usage when Agar windows and container widgets are resized.
- [**AG_Object**](https://libagar.org/man3/AG_Object): In `AG_ObjectDestroy()`, do not call `AG_ObjectDetach()` on child objects. Instead, just cancel any active timers and null their parent pointers without raising a "detached" event. This allows us to perform garbage collection more efficiently.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): Make `AG_TLIST_NO_BGLINES` the default.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): In `AG_TlistNewPolledMs()`, accept a `ms` argument of 0 to disable periodical updates (the caller may still trigger updates with "tlist-poll").
- In `AG_WidgetCopyStyle()`, prevent redundant locking by using `AG_AccessVariable_NoLock()`.

### Fixed
- [**AG_Widget**](https://libagar.org/man3/AG_Widget): Fix redundant "detached" handler invocations. Improve overall efficiency of the garbage-collection step after windows are destroyed.
- [**AG_Console**](https://libagar.org/man3/AG_Console): Prevent a redundant lock operation in the backend of `AG_ConsoleMsg()`.

## [1.7.1] - 2025-11-15
### Added
- An awesome new Agar logo. Thanks [eightbit](https://8bitprodigy.github.io)!
- Added support for the [Cmake](https://cmake.org) build system.
- Added support for the [Meson](https://mesonbuild.com) build system. Thanks [coleman](https://git.sr.ht/~coleman/)!
- New functions [**AG_InitVideoSDL2**](https://libagar.org/man3/AG_InitVideoSDL2) and [**AG_SetVideoSurfaceSDL2**](https://libagar.org/man3/AG_SetVideoSurfaceSDL2) for integrating with an existing SDL2 display context. Thanks Brigham Keys!
- [**AG_FileDlg**](https://libagar.org/man3/AG_FileDlg): Add "Any File" type.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): New function `AG_TlistCopy()`. Copy all items from a source to a destination `AG_Tlist`.
- [**AG_Combo**](https://libagar.org/man3/AG_Combo): New constructor `AG_ComboNewFn()`. Sets a `combo-expanded` handler.
- [**AG_Combo**](https://libagar.org/man3/AG_Combo): New member `nVisItems`. Set the number of items to show by default in expansions.
- [**AG_Checkbox**](https://libagar.org/man3/AG_Checkbox): New functions `AG_CheckboxText()` and `AG_CheckboxTextS()` to update the text label.
- Tweak the default color scheme. Document style defaults as a comment block in `gui/style.css`.
- New Ada bindings for the `Window`, `Box`, `Button`, `Checkbox` and `Combo` widget classes. Updated `ada/gui/demo`.

### Removed
- [**AG_Db**](https://libagar.org/man3/AG_Db): Remove support for DB5.
- [**AG_Window**](https://libagar.org/man3/AG_Window): Remove an unnecessary counter variable in `SizeAllocate()`.

### Changed
- Small cosmetic changes to the default color scheme. Set default font size back to 14pt.
- [**AG_Titlebar**](https://libagar.org/man3/AG_Titlebar): Enable the `AG_BUTTON_EXCL` optimization on titlebar buttons.
- [**AG_Combo**](https://libagar.org/man3/AG_Combo): Allow the dropdown list to be populated by a `combo-expanded` handler.
- [**AG_UCombo**](https://libagar.org/man3/AG_UCombo): Allow the dropdown list to be populated by a `ucombo-expanded` handler.

### Fixed
- [**AG_Combo**](https://libagar.org/man3/AG_Combo): Make it again possible to statically initialize `list` before `combo-expanded`. Restores compatibility pre-1.6. Thanks Wally!
- [**AG_FileDlg**](https://libagar.org/man3/AG_FileDlg): Fix widget geometries not updating when switching to a different Type filter.
- [**AG_FileDlg**](https://libagar.org/man3/AG_FileDlg): Fix compilation where `GLOB_TILDE` is unavailable.
- SDL2 drivers: Require at least version 2.0.22 of SDL2 (`SDL_HINT_MOUSE_AUTO_CAPTURE` is necessary for best user experience).
- [**AG_ProgressBar**](https://libagar.org/man3/AG_ProgressBar): Make `padding` work as expected in progress bar. Thanks scaramacai!
- [**cocoa**](https://libagar.org/man3/AG_DriverCocoa): Fix a compilation error.
- [**glx**](https://libagar.org/man3/AG_DriverGLX): Fix a keyboard handling error. Don't pass `KeyRelease` events to `XwcLookupString()`. Fixes issue #74. Thanks @tsesani and @scaramacai!
- [**glx**](https://libagar.org/man3/AG_DriverGLX): Introduce the `no-im` (disable input methods) driver option.
- [**MAP**](https://libagar.org/man3/MAP): Fix a crash when attempting to create an object if `pLibsRoot` is NULL.
- [**AG_Web**](https://libagar.org/man3/AG_Web): Fix inline compilation problem due to a typo in `net/web.h`.
- On Darwin platforms, build with `_DARWIN_C_SOURCE` globally. Fixes issue #77. Thanks @kwhr0!
- On FreeBSD check for availability of `CLOCK_SECOND` in addition to `HAVE_CLOCK_GETTIME`.
- Test for the `dirfd()` function at compilation.

trunk/engine/mapedit/objq.c
/*	$Csoft: objq.c,v 1.18 2002/09/13 11:08:30 vedge Exp $	*/

/*
 * Copyright (c) 2002 CubeSoft Communications, Inc. <http://www.csoft.org>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistribution of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Neither the name of CubeSoft Communications, nor the names of its
 *    contributors may be used to endorse or promote products derived from
 *    this software without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 * USE OF THIS SOFTWARE EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <sys/types.h>

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include <engine/engine.h>
#include <engine/map.h>

#include <engine/widget/widget.h>
#include <engine/widget/window.h>
#include <engine/widget/primitive.h>
#include <engine/widget/button.h>

#include "mapedit.h"
#include "mapview.h"
#include "objq.h"
#include "fileops.h"

static const struct widget_ops objq_ops = {
	{
		objq_destroy,	/* destroy */
		NULL,		/* load */
		NULL		/* save */
	},
	objq_draw,
	NULL		/* update */
};

static void	 objq_scaled(int, union evarg *);
static void	 objq_event(int, union evarg *);
static void	 tilemap_option(int, union evarg *);

enum {
	SELECTION_COLOR,
	GRID_COLOR
};

enum {
	SELECT_BUTTON =		1,
	SCROLL_BUTTON_MASK =	(SDL_BUTTON_MMASK|SDL_BUTTON_RMASK),
	SCROLL_LEFT_KEY =	SDLK_LEFT,
	SCROLL_RIGHT_KEY =	SDLK_RIGHT
};

struct objq *
objq_new(struct region *reg, struct mapedit *med, int flags, int rw, int rh)
{
	struct objq *objq;

	objq = emalloc(sizeof(struct objq));
	objq_init(objq, med, flags, rw, rh);

	pthread_mutex_lock(&reg->win->lock);
	region_attach(reg, objq);
	pthread_mutex_unlock(&reg->win->lock);

	return (objq);
}

void
objq_init(struct objq *oq, struct mapedit *med, int flags, int rw, int rh)
{
	widget_init(&oq->wid, "objq", "widget", &objq_ops, rw, rh);
	widget_map_color(oq, SELECTION_COLOR, "objq-selection", 0, 200, 0);
	widget_map_color(oq, GRID_COLOR, "objq-grid", 208, 208, 208);
	
	oq->med = med;
	oq->offs = 0;
	oq->flags = (flags != 0) ? flags : OBJQ_HORIZ;
	SLIST_INIT(&oq->tmaps);
	
	event_new(oq, "widget-scaled", 0,
	    objq_scaled, NULL);
	event_new(oq, "window-mousemotion", 0,
	    objq_event, "%i", WINDOW_MOUSEMOTION);
	event_new(oq, "window-mousebuttonup", 0,
	    objq_event, "%i", WINDOW_MOUSEBUTTONUP);
	event_new(oq, "window-mousebuttondown", 0,
	    objq_event, "%i", WINDOW_MOUSEBUTTONDOWN);
}

static void
tilemap_option(int argc, union evarg *argv)
{
	struct mapview *mv = argv[1].p;
	int opt = argv[2].i;

	switch (opt) {
	case MAPEDIT_TOOL_GRID:
		if (mv->flags & MAPVIEW_GRID) {
			mv->flags &= ~(MAPVIEW_GRID);
		} else {
			mv->flags |= MAPVIEW_GRID;
		}
		break;
	case MAPEDIT_TOOL_PROPS:
		if (mv->flags & MAPVIEW_PROPS) {
			mv->flags &= ~(MAPVIEW_PROPS);
		} else {
			mv->flags |= MAPVIEW_PROPS;
		}
		break;
	case MAPEDIT_TOOL_SHOW_CURSOR:
		if (mv->flags & MAPVIEW_SHOW_CURSOR) {
			mv->flags &= ~(MAPVIEW_SHOW_CURSOR);
		} else {
			mv->flags |= MAPVIEW_SHOW_CURSOR;
		}
		break;
	case MAPEDIT_TOOL_EDIT:
		if (mv->flags & MAPVIEW_EDIT) {
			mv->flags &= ~(MAPVIEW_EDIT);
		} else {
			mv->flags |= MAPVIEW_EDIT;
		}
		break;
	}
}

static void
objq_select(struct objq *oq, struct mapedit *med, struct editobj *eob)
{
	static int cury = 85;
	struct object *ob = eob->pobj;
	struct objq_tmap *tm;
	struct window *win;
	struct region *reg;
	struct mapview *mv;
	struct button *bu;
	char *wname;

	SLIST_FOREACH(tm, &oq->tmaps, tmaps) {
		if (tm->ob == eob->pobj) {
			window_show(tm->win);
			view_focus(tm->win);
			return;
		}
	}

	win = emalloc(sizeof(struct window));
	if ((cury += 32) + 264 > view->h) {
		cury = 85;
	}

	asprintf(&wname, "t-win-%s", OBJECT(ob->art->map)->name);
	window_init(win, wname, ob->name, WINDOW_SOLID,
	    view->w - 170, cury, 154, 264, 156, 94);
	free(wname);

	/* Map view */
	mv = emalloc(sizeof(struct mapview));
	mapview_init(mv, med, ob->art->map,
	    MAPVIEW_CENTER|MAPVIEW_ZOOM|MAPVIEW_TILEMAP|MAPVIEW_GRID|
	    MAPVIEW_PROPS|MAPVIEW_SHOW_CURSOR,
	    100, 100);
	
	/*
	 * Tools
	 */
	reg = region_new(win, REGION_HALIGN, 0, 0, 100, -25);
	reg->spacing = 1;

	/* Load map */
	bu = button_new(reg, NULL, SPRITE(med, MAPEDIT_TOOL_LOAD_MAP),
	    0, -1, -1);
	WIDGET(bu)->flags |= WIDGET_NO_FOCUS;
	event_new(bu, "button-pushed", 0, fileops_revert_map, "%p", mv);

	/* Save map */
	bu = button_new(reg, NULL, SPRITE(med, MAPEDIT_TOOL_SAVE_MAP),
	    0, -1, -1);
	WIDGET(bu)->flags |= WIDGET_NO_FOCUS;
	event_new(bu, "button-pushed", 0, fileops_save_map, "%p", mv);

	/* Grid */
	bu = button_new(reg, NULL, SPRITE(med, MAPEDIT_TOOL_GRID),
	    BUTTON_STICKY|BUTTON_PRESSED, -1, -1);
	WIDGET(bu)->flags |= WIDGET_NO_FOCUS;
	event_new(bu, "button-pushed", 0,
	    tilemap_option, "%p, %i", mv, MAPEDIT_TOOL_GRID);

	/* Props */
	bu = button_new(reg, NULL, SPRITE(med, MAPEDIT_TOOL_PROPS),
	    BUTTON_STICKY|BUTTON_PRESSED, -1, -1);
	WIDGET(bu)->flags |= WIDGET_NO_FOCUS;
	event_new(bu, "button-pushed", 0,
	    tilemap_option, "%p, %i", mv, MAPEDIT_TOOL_PROPS);
	
	/* Edition */
	bu = button_new(reg, NULL, SPRITE(med, MAPEDIT_TOOL_EDIT),
	    BUTTON_STICKY, -1, -1);
	WIDGET(bu)->flags |= WIDGET_NO_FOCUS;
	event_new(bu, "button-pushed", 0,
	    tilemap_option, "%p, %i", mv, MAPEDIT_TOOL_EDIT);

	/* Map view */
	reg = region_new(win, REGION_HALIGN, 0, 10, 100, 90);
	region_attach(reg, mv);

	tm = emalloc(sizeof(struct objq_tmap));
	tm->win = win;
	tm->ob = ob;
	SLIST_INSERT_HEAD(&oq->tmaps, tm, tmaps);

	view_attach(win);
	window_show(win);
}

static void
objq_event(int argc, union evarg *argv)
{
	struct objq *oq = argv[0].p;
	struct mapedit *med = oq->med;
	struct editobj *eob;
	int button, x, ox, ms;
	int curoffs;

	switch (argv[1].i) {
	case WINDOW_MOUSEMOTION:
		x = argv[2].i / TILEW;
		ox = oq->mouse.x;
		oq->mouse.x = x;

		ms = SDL_GetMouseState(NULL, NULL);
		if (ms & SCROLL_BUTTON_MASK) {
			if (oq->mouse.x > ox &&		/* Down */
			    --oq->offs < 0) {
				oq->offs = med->neobjs - 1;
			}
			if (oq->mouse.x < ox &&		/* Up */
			    ++oq->offs > med->neobjs-1) {
				oq->offs = 0;
			}
		}
		break;
	case WINDOW_MOUSEBUTTONUP:
	case WINDOW_MOUSEBUTTONDOWN:
		button = argv[2].i;
		x = argv[3].i;
		if (button != SELECT_BUTTON) {
			break;
		}
		curoffs = oq->offs + x/TILEW;
		if (curoffs < 0) {
			/* Wrap */
			curoffs += oq->offs + med->neobjs;
		} else if (med->ref.offs >= med->neobjs) {
			curoffs -= med->neobjs;
		}
		while (curoffs > med->neobjs-1) {
			curoffs -= med->neobjs;
		}
		TAILQ_INDEX(eob, &med->eobjsh, eobjs, curoffs);
		objq_select(oq, med, eob);
		break;
	}
}

static void
objq_scaled(int argc, union evarg *argv)
{
	struct objq *oq = argv[0].p;
	int maxw = argv[1].i, maxh = argv[2].i;
	int w, h;

	for (w = 0; w < WIDGET(oq)->w-1 && w < maxw; w += TILEW) ;;
	for (h = 0; h < WIDGET(oq)->h-1 && h < maxh; h += TILEH) ;;
	
	WIDGET(oq)->w = w;
	WIDGET(oq)->h = h;
}

void
objq_destroy(void *p)
{
	struct objq *oq = p;
	struct objq_tmap *tm, *ntm;

	for (tm = SLIST_FIRST(&oq->tmaps);
	     tm != SLIST_END(&oq->tmaps);
	     tm = ntm) {
		ntm = SLIST_NEXT(tm, tmaps);
		free(tm);
	}

	widget_destroy(p);
}

void
objq_draw(void *p)
{
	struct objq *oq = p;
	struct mapedit *med = oq->med;
	struct editobj *eob;
	int x = 0, i, sn;
	
	for (i = 0, sn = oq->offs;
	     i < (WIDGET(oq)->w / TILEW);
	     i++, x += TILEW) {
		if (sn > -1) {
			/* Absolute */
			TAILQ_INDEX(eob, &med->eobjsh, eobjs, sn);
		} else {
			/* Wrap */
			TAILQ_INDEX(eob, &med->eobjsh, eobjs,
			    sn + med->neobjs);
			if (eob == NULL) {
				goto nextref;
			}
		}

		WIDGET_DRAW(oq, SPRITE(eob->pobj, 0), x, 0);

		if (med->curobj == eob) {
			primitives.square(oq, x, 0, TILEW, TILEH,
			    WIDGET_COLOR(oq, SELECTION_COLOR));
			primitives.square(oq, x+1, 1, TILEW-1, TILEH-1,
			    WIDGET_COLOR(oq, SELECTION_COLOR));
		} else {
			primitives.square(oq, x, 0, TILEW, TILEH,
			    WIDGET_COLOR(oq, GRID_COLOR));
		}
nextref:
		if (++sn >= med->neobjs) {
			sn = 0;
		}
	}
}

